### PR TITLE
Feature #40 merge http ws ports

### PR DESCRIPTION
**Pull request by ballinette**
merge websocket, socketIO and http protocols on same port (websocket and socketIO use now the HTTP server)
It makes the `kuzzle-plugin-websocket` and `kuzzle-plugin-socketio` plugins obsolete

+ add settings to enable/disable core protocols (by default, both 3 are enabled).

Note : if `http` protocol is disabled, the http server will still be started in order to run websocket and/or socketio protocols, but no HTTP route will be available.

TODO after merging this PR :  "un-wip" the related PRs on other repositories:
* https://github.com/kuzzleio/sdk-javascript/pull/180
* https://github.com/kuzzleio/documentation/pull/86
* https://github.com/kuzzleio/kuzzle/pull/666

### CODE REVIEW

File: lib/service/HttpProxy.js [hunk 1]
```diff
@@ -48,97 +48,101 @@ HttpProxy.prototype.init = function (proxy) {
...
-    const connection = new ClientConnection('HTTP/' + request.httpVersion, ips, request.headers);
-    _proxy.clientConnectionStore.add(connection);
-
-    if (request.headers['content-length'] > this.maxRequestSize) {
-      request
-        .removeAllListeners()
-        .resume();
-      return replyWithError(connection.id, payload, response, new SizeLimitError('Error: maximum HTTP request size exceeded'));
-    }
+  if (proxy.config.http.enabled) {
```
Comment (benoitvidis) [inline lib/service/HttpProxy.js]: nitpicking: would be easier to read if we would first test if the http protocol is disabled, and if so, create the blank server and exit.

File: lib/service/protocol/SocketIo.js [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+ */
+function SocketIo () {
+  this.sockets = {};
+  this.io = null;
+
+  this.init = function (proxy) {
+    _proxy = proxy;
+
+    if (_proxy.config.socketio.enabled) {
+      this.io = require('socket.io')(proxy.httpProxy.server);
```
Comment (benoitvidis) [inline lib/service/protocol/SocketIo.js]: nitpicking: would deserve a comment line imo stating the path defaults to `/socket.io` and we don't need to specify it (http://socket.io/docs/server-api/#server#path(v:string):server)

File: lib/service/protocol/SocketIo.js [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+ * @constructor
+ */
+function SocketIo () {
+  this.sockets = {};
+  this.io = null;
+
+  this.init = function (proxy) {
+    _proxy = proxy;
+
+    if (_proxy.config.socketio.enabled) {
```
Comment (benoitvidis) [inline lib/service/protocol/SocketIo.js]: same as for http: early exit on the negative predicate is more readable.

File: lib/core/config.js [hunk 4]
```diff
@@ -58,9 +58,16 @@ config = rc('proxy', {
     timeout: 10000
   },
   http: {
+    enabled: true,
     port: 7511,
```
Comment (benoitvidis) [inline lib/core/config.js]: I would suggest to switch back to 7512, which was has a little story about how it was chosen.

File: lib/service/protocol/SocketIo.js [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+      this.io = require('socket.io')(proxy.httpProxy.server);
+      this.io.set('origins', '*:*');
```
Comment (stafyniaksacha) [inline lib/service/protocol/SocketIo.js]: Can we have this in settings ?

Comment (ballinette) [inline lib/service/protocol/SocketIo.js]: Yes, we can ;)
I just reported what was actually done in the plugin socketio, but you're right ;)

File: lib/core/config.js (hunk 4 above)
Comment (ballinette) [inline lib/core/config.js]: done

File: lib/service/HttpProxy.js (hunk 1 above)
Comment (ballinette) [inline lib/service/HttpProxy.js]: done

File: lib/service/protocol/SocketIo.js (hunk 3 above)
Comment (ballinette) [inline lib/service/protocol/SocketIo.js]: done

File: lib/service/protocol/SocketIo.js (hunk 2 above)
Comment (ballinette) [inline lib/service/protocol/SocketIo.js]: done

File: lib/service/protocol/SocketIo.js (hunk 5 above)
Comment (ballinette) [inline lib/service/protocol/SocketIo.js]: done

**Comment by ballinette** [comment]
@sstafyniak @benoitvidis all your remarks have been treated

**Comment by ballinette** [comment]
Documentation's PR created: https://github.com/kuzzleio/documentation/pull/86

**Pull request merged**